src/key_value_store.rs

```diff
@@ -0,0 +1,74 @@
+use crate::persistence::EncryptingResourceAdapter;
```

Comment:
It looks like this is editing the in-memory hashmap, and then storing it as a snapshot just before doing a commit. It's not quite the same interface.

The KVStore in atomicstore is build on a serialized `AppendLog<ResourceAdaptor<<key, Option<value>>>`, with an optional in-memory `HashMap<key, StorageLocation>` or fully loaded with `HashMap<key, Value>`; the intent is to have the in-memory part be fully optional once the native filesystem representation is implemented.

The supported operations are
`fn store_resource(key, value) -> Result<StorageLocation>`
`fn clear_resource(key) -> Result<bool>` // true if key existed
`fn load_resource(key) -> Result<Option<value>>`

There's an iterator, but it isn't exposed.

The store is assumed to overwrite if a previous value existed for the key.

Both `store_resource` and `clear_resource` immediately call the `AppendLog::store_resource(key, Some(value))` or `(key, None)` and `commit_version` and `revert_version` get forwarded. The overhead in the current implementation is in building the `HashMap` on initial load, and updating it for the `clear_resource` and `store_resource` calls.

Reply:
That all sounds reasonable (and having the in-memory index be fully optional is awesome!)

Why isn't the iterator exposed? That seems like something we would want to support.

The disadvantage of storing it as an append log, rather than snapshots, is that for collections which change frequently, the append log can grow much larger than the size of the actual collection. Is there any plan for pruning outdated entries?

Reply:
The iterator has too much knowledge about the underlying implementation, and won't really work the same once we replace the implementation.

The plan is for the append log implementation to be a stopgap. The upcoming implementation will have directories corresponding to hash buckets on the keys, with files tagged with version storing key+value, and options to retain (which allows the hash table to be queried at any specified historical version) or purge (which allows lower space utilization) after commit.

The advantage of a rolling log is that the size is relative to the total number of current keys, times the number of entries stored, but the actual size is still going to be larger than the append log until something in excess of number of entries times as many keys have been updated to a new value as the total number of keys.

e.g., if there are 20,000 keys, and on average any given key has been updated 10 times, and the rolling log retention is 10, the append log approach will be around the same size... but eventually it will be bigger, once the average number of updates gets significantly larger. With frequent deletions, the disparity is more pronounced. However, the rolling log writes will take the better part of forever with that many keys.

In a scenario with only 100 keys, but values changing frequently, the rolling log is a very clear win.

However, I figured the odds were good that we wouldn't reach the point where we had a scenario where a rolling log was substantially better before we had a real KVStore to use.

Reply:
Eventually having a stable iterator interface seems very important. But for now I suppose we can just get the iterator from the in-memory index.

Reply:
I changed the `RollingLog` to `AppendLog` and modified the `store` and `delete` functions to be consistent with the atomicstore. I didn't change `index` to optional though, since I think it's still needed for some `KeyValueStore` functions. Let me know if this seems off. Thanks!